### PR TITLE
Migrate docs to docs.rs and fix backend docs gen

### PR DESCRIPTION
**Pull request by pinkforest**
I've tested backend doc generation here (Some links point to the real crate which needed absolute refs)
https://docs.rs/test-curve25519-dalek

```bash
cargo rustdoc --features "simd_backend packed_simd" -- \
 --html-in-header docs/assets/rustdoc-include-katex-header.html \
 --cfg docsrs
```

docsrs feature documents the necessary things in docs.rs - now backend modules get documented

docs.rs does not support documenting pub(crate) so I had to work/around this with:

```rust
// Arithmetic backends (using u32, u64, etc) live here
#[cfg(docsrs)]
pub mod backend;
#[cfg(not(docsrs))]
pub(crate) mod backend;
```

Details: https://docs.rs/project-template/latest/project_template/

Above is needed since some of the required katex renders comes from there.

Some markdown links require absolute paths e.g.

```
[parallel_doc]: https://docs.rs/curve25519-dalek/latest/curve25519_dalek/backend/vector/index.html
```

As this is both used from repo and docs.rs and there is no way to link in docs.rs relatively as docs.rs always includes the crate/version/crate_normalised
docs.rs ensured the katex'ed markdown is rendered correctly into HTML

Also local documentation can be generated as per instructions w/ cargo rustdoc via make doc

I recommend creating a pre-release to ensure all the docs work and see those docs get rendered ok esp on backend even if I have done for testing already in docs.rs by using different crate name.

Using https://github.com/lycheeverse/lychee to check links - TBC

As expected backend/vector is not yet in docs.rs as this is not published yet there

$ lychee README.md 
```
⠁ 6/17 [00:00:00] ████████░░░░░░░░░░░░░░░░░ ✔ [200] https://img.shields.io/badge/dynamic/json.svg?label=docs&uri=https%3A%2F%2Fcrates.io%2Fapi%2Fv1%2Fcrates%2Fcurve25519-dalek%2Fversions&query=%24.versions%5B0%5D.numIssues found in 1 input. Find details below.

[README.md]:
✗ [404] https://docs.rs/curve25519-dalek/latest/curve25519_dalek/backend/vector/index.html | Failed: Network error: Not Found

🔍 17 Total ✅ 16 OK 🚫 1 Error (HTTP:1)
```

The redirect error below can be ignored - lychee red herring

$ lychee docs/
```
⠤ 15/15 [00:00:08] █████████████████████████ ✗ [ERR] https://software.intel.com/sites/default/files/managed/9e/bc/64-ia-32-architectures-optimization-manual.pdf | Failed: Network error: error following redirect for uIssues found in 1 input. Find details below.

[docs/parallel-formulas.md]:
✗ [ERR] https://software.intel.com/sites/default/files/managed/9e/bc/64-ia-32-architectures-optimization-manual.pdf | Failed: Network error: error following redirect for url (https://software.intel.com/content/www/us/en/develop/download/intel-64-and-ia-32-architectures-optimization-reference-manual.html): too many redirects

🔍 15 Total ✅ 14 OK 🚫 1 Error (HTTP:1)
```

Even the generated doc was fine - all the errors were safe to ignore just deps / red herrings and the above disclaimer getting it published first :)

$ lychee target/doc/
```
🔍 452146 Total ✅ 452081 OK 🚫 65 Errors (HTTP:65)
```

### CODE REVIEW

**Comment by jplatte** [comment]
> docs.rs does not support documenting pub(crate)

* Does `cargo-args = ["--document-private-items"]` or `rustdoc-args = ["--document-private-items"]` not work?
* Why should a module be documented that isn't part of the public API?

**Comment by pinkforest** [comment]
TL;DR The katex-markdown rendered HTML's behind backend (e.g. vector) are leaked already in the public side and when the public side does not have working link it becomes a nightmare to maintain them as the code changes.

mod backend is a special gatekeeper on how the code / documentation was structured previously that triggers whole lot katex'es markdown to html generation behind it as these are associated to

With the current structure working with docs.rs it's either that or self-hosting the site with private / public wholesale binary divide - as before - that docsrs does not support where we don't want to document private items otherwise other than the katex ones.

docs.rs has open issue here:
https://github.com/rust-lang/docs.rs/issues/304

The code structure might be changed later for documentation but for 4.0.0 release we needed something that still sounds reasonable.

However I considered that we could document the private items in github that includes the katex's but then the public side is still linking to these private items which becomes a pain to maintain as they are part of the code.

You can already see this when actually building the documentation as there are warnings when you generate the public documentation that e.g. vector backend is missing and there is no seamless linking between public / private items where there is a leakage from public to private side.

File: src/backend/mod.rs [hunk 1]
```diff
@@ -36,18 +36,5 @@
...
-    all(feature = "nightly", rustdoc)
-))]
-#[cfg_attr(
-    feature = "nightly",
-    doc(cfg(any(all(
-        feature = "simd_backend",
-        any(target_feature = "avx2", target_feature = "avx512ifma")
-    ))))
-)]
+#[cfg(any(feature = "simd_backend", docsrs))]
```
Comment (pinkforest) [inline src/backend/mod.rs:39]: If used with stable for some reaosn this will now do a compiler error instead of failing silently.

File: src/backend/vector/mod.rs [hunk 2]
```diff
@@ -11,29 +11,29 @@
...
-#[cfg(not(any(target_feature = "avx2", target_feature = "avx512ifma", rustdoc)))]
+#[cfg(not(any(target_feature = "avx2", target_feature = "avx512ifma", docsrs)))]
 compile_error!("simd_backend selected without target_feature=+avx2 or +avx512ifma");
 
 #[cfg(any(
     all(target_feature = "avx2", not(target_feature = "avx512ifma")),
-    rustdoc
+    docsrs
 ))]
-#[doc(cfg(all(target_feature = "avx2", not(target_feature = "avx512ifma"))))]
```
Comment (pinkforest) [inline src/backend/vector/mod.rs:21]: doc-cfg is annoying unstable nightly feature and we use cfg's for this same thing already.

File: Cargo.toml [hunk 3]
```diff
@@ -24,9 +24,8 @@ exclude = [
 ]
 
 [package.metadata.docs.rs]
-# Disabled for now since this is borked; tracking https://github.com/rust-lang/docs.rs/issues/302
-# rustdoc-args = ["--html-in-header", ".cargo/registry/src/github.com-1ecc6299db9ec823/curve25519-dalek-0.13.2/rustdoc-include-katex-header.html"]
-features = ["nightly", "simd_backend"]
+rustdoc-args = ["--html-in-header", "docs/assets/rustdoc-include-katex-header.html", "--cfg", "docsrs"]
```
Comment (pinkforest) [inline Cargo.toml:27]: rustdoc feature is no more available for cfg - Replaced with docsrs for docs.rs specific generation everyone else uses.

**Review by rozbb — CHANGES_REQUESTED** [review]

File: src/scalar.rs [hunk 4]
```diff
@@ -401,6 +411,7 @@ use serde::de::Visitor;
 use serde::{self, Deserialize, Deserializer, Serialize, Serializer};
 
 #[cfg(feature = "serde")]
+#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
```
Comment (jplatte) [inline src/scalar.rs:414]: If you add `#![cfg_attr(docsrs, feature(doc_auto_cfg))]` to the crate attributes, `doc(cfg)`s that exactly mirror regular `cfg`s at the same level will be unnecessary.

Comment (rozbb) [reply src/scalar.rs:414]: Unfortunately it leaves an "or `docsrs`" note on all the things in `backend`. So I did it manually 🤷‍♂️

File: Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (pinkforest) [reply Cargo.toml:27]: Everyone else uses docrs so it aligns to that

[the remark this answers is not in the document]
Comment (pinkforest) [reply Cargo.toml:27]: We can also add also remove the flag from internal docs generation

[the remark this answers is not in the document]
Comment (rozbb) [reply Cargo.toml:27]: Good enough for me

[the remark this answers is not in the document]
Comment (rozbb) [reply Cargo.toml:27]: > We can also add also remove the flag from internal docs generation

 what do you mean?

[the remark this answers is not in the document]
Comment (pinkforest) [reply Cargo.toml:27]: Locally we don't need to use docsrs flag as backend really is implementation detail by it's academic intepretation
We only need it for the docs.rs public consumption where we need to present consumable public story.

File: docs/parallel-formulas.md [hunk 5]
```diff
@@ -327,7 +327,7 @@ There are several directions for future improvement:
 [sandy2x]: https://eprint.iacr.org/2015/943.pdf
 [avx2trac]: https://trac.torproject.org/projects/tor/ticket/8897#comment:28
 [hwcd08]: https://www.iacr.org/archive/asiacrypt2008/53500329/53500329.pdf
-[curve_models]: https://doc-internal.dalek.rs/curve25519_dalek/backend/serial/curve_models/index.html
+[curve_models]: https://docs.rs/dalek-test-curve-docs/latest/dalek_test_curve_docs/backend/serial/curve_models/index.html
```
[the remark this answers is not in the document]
Comment (pinkforest) [reply docs/parallel-formulas.md:330]: Can't use relative as docs.rs requires the full path and it has to be linked also from local documentation which can be different.. it's a pain I know

File: Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (rozbb) [reply Cargo.toml:27]: Ah I see. Yeah, makes sense

File: docs/parallel-formulas.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (rozbb) [reply docs/parallel-formulas.md:330]: No I mean it's `dalek-test-curve-docs` rather than `curve25519-dalek`

[the remark this answers is not in the document]
Comment (pinkforest) [reply docs/parallel-formulas.md:330]: ooops good catch :heart:

File: src/scalar.rs (hunk 4 above)
Comment (rozbb) [reply src/scalar.rs:414]: Hmm, maybe I can do `doc_auto_cfg` and then just do manual things for the `backend` parts

**Comment by rozbb** [comment]
This looks great! I left a few notes. Two things:

1. I agree that `backend` should be the only non-public thing that's documented as if it were public. Would it be appropriate to add `**INTERNALS:**` or something like that in the first line of the module docs?
2. The current doc build scripts don't work on non-x86_64 arches (it fails on the `vector` backend, since it requires some `std::arch::x86_64` items). Is there a cfg flag that will exclude it if the machine isn't x86_64?

cc @tarcieri above and also I want your opinion on is whether using docs.rs and eschewing the entire idea of self-hosting docs is appropriate. I'm pretty strongly in favor. And if people want truly internal docs (not just `backend`plus the public stuff) they can compile locally.

File: docs/parallel-formulas.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (pinkforest) [reply docs/parallel-formulas.md]: Fixed :+1:

**Comment by pinkforest** [comment]
> 1.  agree that backend should be the only non-public thing that's documented as if it were public. Would it be appropriate to add **INTERNALS:** or something like that in the first line of the module docs?

Yeah I'll fix.

> 2. The current doc build scripts don't work on non-x86_64 arches (it fails on the vector backend, since it requires some std::arch::x86_64 items). Is there a cfg flag that will exclude it if the machine isn't x86_64?

`not(target_arch = "x84_64")` but it has to be probably `target_arch = "x86_64"` as it's "inclusive" compilation.

File: src/scalar.rs (hunk 4 above)
Comment (jplatte) [reply src/scalar.rs:414]: > Unfortunately it leaves an "or `docsrs`" note on all the things in `backend`. So I did it manually :man_shrugging:

I think that should be fixable through `#![cfg_attr(docsrs, feature(doc_cfg_hide))]` + `#![doc(cfg_hide(docsrs))]`.

**Comment by pinkforest** [comment]
You can get rid of the local compilation on non-x86_64 by removing the docsrs flag
docs.rs is x86_64 so it works there but all we need to adjust for it is the make docs and remove docsrs from it

File: src/scalar.rs (hunk 4 above)
Comment (rozbb) [reply src/scalar.rs:414]: Fixed. Thanks for the idea!

**Comment by rozbb** [comment]
Ok, maybe we can punt on it for now. I think this is just about done! I g2g sleep. I accidentally pulled an all-nighter

**Review by rozbb — APPROVED** [review]

**Comment by pinkforest** [comment]
Re-rendered: https://docs.rs/crate/test-curve25519-dalek/4.0.0-pre.8

**Comment by rozbb** [comment]
Oh also, I have no idea why the dalek logo is now above the main text and not set within it. That's been like that for a while

**Comment by pinkforest** [comment]
docs.rs has it's own style thing, can do nothing about that.

File: src/scalar.rs (hunk 4 above)
Comment (jplatte) [reply src/scalar.rs:414]: I don't actually see where you enabled it in the commit that says you did, but happy to help 😄

**Comment by pinkforest** [comment]
@rozbb

https://docs.rs/test-curve25519-dalek/4.0.0-pre.8 is the current this PR

https://docs.rs/test-curve25519-dalek/4.0.0-pre.9/test_curve25519_dalek/#modules is below diff if you want to try out:
```
 git diff src/backend/vector/mod.rs
diff --git a/src/backend/vector/mod.rs b/src/backend/vector/mod.rs
index 29a188f..734c442 100644
--- a/src/backend/vector/mod.rs
+++ b/src/backend/vector/mod.rs
@@ -16,24 +16,24 @@ compile_error!("simd_backend selected without target_feature=+avx2 or +avx512ifm
 
 #[cfg(any(
     all(target_feature = "avx2", not(target_feature = "avx512ifma")),
-    docsrs
+    all(docsrs, target_arch = "x86_64")
 ))]
 pub mod avx2;
 #[cfg(any(
     all(target_feature = "avx2", not(target_feature = "avx512ifma")),
-    docsrs
+    all(docsrs, target_arch = "x86_64")
 ))]
 pub(crate) use self::avx2::{
     constants::BASEPOINT_ODD_LOOKUP_TABLE, edwards::CachedPoint, edwards::ExtendedPoint,
 };
 
-#[cfg(any(target_feature = "avx512ifma", docsrs))]
+#[cfg(any(target_feature = "avx512ifma", all(docsrs, target_arch = "x86_64")))]
 pub mod ifma;
 #[cfg(target_feature = "avx512ifma")]
 pub(crate) use self::ifma::{
     constants::BASEPOINT_ODD_LOOKUP_TABLE, edwards::CachedPoint, edwards::ExtendedPoint,
 };
 
-#[cfg(any(target_feature = "avx2", target_feature = "avx512ifma", docsrs))]
+#[cfg(any(target_feature = "avx2", target_feature = "avx512ifma", all(docsrs, target_arch = "x86_64")))]
 #[allow(missing_docs)]
 pub mod scalar_mul;
```

**Review by tarcieri — APPROVED** [review]

**Comment by rozbb** [comment]
Thank you again! Gonna do some README updates next.

**Comment by pinkforest** [comment]
Wow it aligns! :partying_face: 
Rendered:
https://docs.rs/test-curve25519-dalek/4.0.0-pre.10/test_curve25519_dalek/

**Comment by rozbb** [comment]
Oh oops I didn't realize it was merged to the wrong branch. Will merge to `release/4.0`

**Comment by pinkforest** [comment]
It would be nice if everyone works via fork's as the branches are getting plenty :)
